akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterShardingImpl.scala

```diff
@@ -256,6 +259,7 @@ import akka.util.JavaDurationConverters._
   override def entityRefFor[M](typeKey: scaladsl.EntityTypeKey[M], entityId: String): scaladsl.EntityRef[M] = {
     new EntityRefImpl[M](
       classicSharding.shardRegion(typeKey.name),
+      useShardingEnvelope.get(typeKey.name),
```

Comment:
There is one thing that I would like to discuss. When creating the EntityRef without envelope the entityId has to be defined twice. In the message itself, and in this entityRefFor parameter. In that case the entityRefFor parameter is never used.

Should we have another entityRefFor that doesn't take the entityId, so it is a EntityRef that can be used for sending to any entity, since the entityId is in the messages. That would only be allowed for NoEnvelopeShardingMessageExtractor (which can only be a runtime check). In that case it's probably good to give it a clear, separate, name, such as `noEnvelopeEntityRefFor`.

Not sure if that would be confusing on the conceptual level, where ActorRef and EntityRef are references to one specific actor/entity individual?

Reply:
Exactly, the `ActorRef[M]` for the shard region is what is intended to be used when not using envelopes. We could have an easier accessor for that. It's currently only returned from `init`.

However, that doesn't solve our original problem of being able to send messages to entities without knowing if they are using ShardingEnvelope or not. Well, we might be able to know and can make our own abstraction in RES gRPC. I'll take a closer look there instead.